---
title: "Integrative meta-analysis of epigenome-wide association studies identifies genomic and epigenomics differences in the brain and the blood in Alzheimerâ€™s disease - Stratified by Sex"
subtitle: 'AIBL dataset'
author:
  - Tiago Chedraoui Silva^[University of Miami]
  - Wei Zhang^[University of Miami]
  - Lily Wang^[University of Miami]
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  rmarkdown::html_document:
    highlight: breezedark
    theme: lumen
    toc: true
    number_sections: true
    df_print: paged
    code_download: false
    toc_float:
      collapsed: yes
    toc_depth: 3
editor_options:
  chunk_output_type: inline    
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,warning = FALSE)
knitr::opts_knit$set(root.dir = "./../../")
```

# Setup

```{R, message = FALSE, results = 'hide'}
library(dplyr)
library(SummarizedExperiment)
library(ExperimentHub)
library(minfi)
library(wateRmelon)
cohort <- "AIBL"
dir.base <- file.path('~/TBL Dropbox/Wei Zhang/')
dir.base.data <- file.path(dir.base,"/AD-meta-analysis-blood-samples/")
dir.base.data2 <- file.path(dir.base,"/AD-meta-analysis-blood-samples-bySex/")
dir.data <- file.path(dir.base.data,"datasets/",cohort,"/") 
dir.data.aux <- file.path(dir.base.data,"datasets/Aux/") 
dir.data.raw <- file.path(dir.data,"/step1_download/") 
dir.data.processed0 <- file.path(dir.data,"/GSE153712_processed/") 
dir.data.processed <- file.path(dir.data,"/GSE153712_processed/withSmoke/") 
dir.data.pca <- file.path(dir.data,"/pca_filtering/withSmoke") 
dir.results <- file.path(dir.base.data2,"analysis_results",cohort,"/single_cpg/withSmoke")
dir.results.glm <- file.path(dir.results,"/Logistic_regression_model")
for(p in grep("dir",ls(),value = T)) dir.create(get(p),recursive = TRUE,showWarnings = FALSE)
```


```{R, eval = TRUE, include = FALSE}
library(GEOquery)
Sys.setenv("VROOM_CONNECTION_SIZE" = 13107200)
query <- getGEO(GEO = "GSE153712")
info <- query$GSE153712_series_matrix.txt.gz
samples.info <- as(phenoData(info),"data.frame")
samples.info$sample <- substr(samples.info$title,1,19)
samples.info <- samples.info[samples.info$`disease status:ch1` %in% c(
  "Alzheimer's disease",
  "healthy control"
),]

nb.samples.male <- sum(samples.info$`gender:ch1` == "Male")
nb.samples.female <- sum(samples.info$`gender:ch1` == "Female")

tab.age <- readr::read_csv(file = file.path(dir.data,"predicted_age.csv"),col_types = readr::cols()) %>%
  dplyr::filter(Elastic_Net >= 65) %>% dplyr::mutate(sample = substr(sample,1,10))
nb.samples.min.age.65.male <- sum(samples.info[samples.info$`gender:ch1` == "Male",]$geo_accession %in% tab.age$sample)
nb.samples.min.age.65.female <- sum(samples.info[samples.info$`gender:ch1` == "Female",]$geo_accession %in% tab.age$sample)
```

# Data retrieval - autosome

```{r, eval = TRUE}
aibl.se <- readRDS(
  file.path(dir.data.pca, "AIBL_QNBMIQ_PCfiltered_age_at_least_65.RDS")
)
```

#  Data Pre-processing - sex chromossomes


```{R, include = FALSE}
RGSet.filtered <- readRDS(file.path(dir.data.processed0,"RGSet_filtered_bs_min_85_min_age_elastic_net_65.rds"))
colnames(RGSet.filtered) <- substr(colnames(RGSet.filtered),1,10 )
RGSet.filtered <- RGSet.filtered[,colnames(aibl.se)]
```


## Limit DNA methylation matrix on chrX & chrY

```{R}
# Find which chromosome each probe is on
annotation <- getAnnotation(RGSet.filtered) %>% as.data.frame %>% dplyr::filter(chr %in% c("chrX","chrY"))
betaSet <- getBeta(RGSet.filtered)
betaSet <- betaSet[rownames(betaSet) %in% rownames(annotation), ]
nb.chrXY.probes <- nrow(betaSet)
```


## Probes QC

1. keep only probes that start with "cg"
2. drop probes where SNP with MAF >= 0.01 was present in the last 5 bp of the probe.


```{R DMRcate, eval = TRUE, inclue = TRUE}
library(DMRcate)
library(minfi)
load(file.path(dir.data.processed0,"detectionP_passed_probes.rda"))
nb.probes.before.norm <- nrow(failed.01)
```

```{R, eval = FALSE}
betaSet <- betaSet[rownames(betaSet) %in% passedProbes, ] # dim 18331 584
nb.chrXY.probes.detectP <- nrow(betaSet)
betaSet <- betaSet[grep("cg",rownames(betaSet)),] # dim 18262
nb.chrX.probes.detectP.cg <- nrow(betaSet)

betaSet <- rmSNPandCH(
  object = betaSet,
  dist = 5,
  mafcut = 0.01,
  and = TRUE,
  rmcrosshyb = TRUE,
  rmXY = FALSE
)
nb.chrXY.probes.detectP.cg.dmrcate <- nrow(betaSet) # 16763
```


## Merge DNA methylation matrix on autosome with the one on chrX, chrY

```{R}
identical(colnames(aibl.se), colnames(betaSet))
betaQN_BMIQ <- rbind(assay(aibl.se), betaSet)
table(base::duplicated(row.names(betaQN_BMIQ))) #777919
```

```{eval = FALSE}
save(
  betaQN_BMIQ,
  file = file.path(dir.data.processed,"betaQN_age_at_least_65_with_XY.rda")
)
```

```{R, include = FALSE, eval = F}
load(file.path(dir.data.processed,"betaQN_age_at_least_65_with_XY.rda"))
```

### Summarize Experiment

```{R, eval = FALSE}
colnames(betaQN_BMIQ) <- substr(colnames(betaQN_BMIQ),1,10 )
betaQN_BMIQ <- betaQN_BMIQ[
  rownames(betaQN_BMIQ) %in% rownames(IlluminaHumanMethylationEPICanno.ilm10b4.hg19::Locations),
]

assay <- S4Vectors::SimpleList("DNAm" = betaQN_BMIQ)

library(GEOquery)

samples.info <- colData(aibl.se)


rowData <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19::Locations[rownames(betaQN_BMIQ),] %>%
  GenomicRanges::makeGRangesFromDataFrame(start.field = "pos",end.field = "pos")
samples.info$PLATE <- factor(samples.info$PLATE)

aibl.se.with.sex <- SummarizedExperiment::SummarizedExperiment(
  assays = assay,
  rowData = rowData,
  metadata = list("genome" = "hg19"),
  colData = samples.info
)

save(aibl.se.with.sex, file = file.path(dir.data.processed,"AIBL_se_age_at_least_65_with_XY.rda"))
readr::write_csv(
  samples.info %>% as.data.frame, 
  file = file.path(dir.data.processed,"AIBL_age_at_least_65_samples_with_XY.info.csv")
)
```


# Summary after QC steps

## Data and metadata
```{R}
aibl.se.with.sex <- get(load(
  file.path(dir.data.processed, "AIBL_se_age_at_least_65_with_XY.rda")
))
aibl.se.with.sex <- aibl.se.with.sex[,aibl.se.with.sex$`disease status:ch1` %in% c("healthy control","Alzheimer's disease")]
nb.samples.after.pca.male <- sum(aibl.se.with.sex$Sex == "M")
nb.samples.after.pca.female <- sum(aibl.se.with.sex$Sex == "F")
nb.probes.after.norm <- nrow(aibl.se.with.sex)
```

```{R, echo = FALSE}
pheno <- colData(aibl.se.with.sex) 
pheno <- pheno %>% as.data.frame %>% dplyr::select(
  c(disease.status.ch1, sample, Sex, age.pred.Elastic_Net, PLATE,
    B, NK,   CD4T, CD8T,   Mono, Neutro, Eosino
  )
)

pheno %>% str()

pheno %>%
  DT::datatable(
    filter = 'top',
    style = "bootstrap",
    extensions = 'Buttons',
    options = list(
      scrollX = TRUE, 
      dom = 'Bfrtip',
      buttons = I('colvis'),
      keys = TRUE, 
      pageLength = 10
    ), 
    rownames = FALSE,
    caption = "Samples metadata"
  )
```



## Number of samples removed in each step

```{R load_plot_pca}
aibl.se.before.pca <- get(load(file.path(dir.data.processed,"AIBL_se_age_at_least_65.rda")))
### transform to m values
pheno_df <- colData(aibl.se.before.pca) %>% as.data.frame
pheno_df <- pheno_df[pheno_df$disease.status.ch1 %in% c("healthy control","Alzheimer's disease"),]
nb.samples.before.pca.male <- sum(pheno_df$Sex == "M")
nb.samples.before.pca.female <- sum(pheno_df$Sex == "F")
```

## Male
```{R, eval = TRUE}
df.samples <- data.frame(
  "Number of samples" =  c(
    nb.samples.male, 
    nb.samples.male, 
    nb.samples.min.age.65.male,
    nb.samples.before.pca.male, 
    nb.samples.after.pca.male
    
  ),
  "Description" = c(
    "Number of samples before bisulfiteConversion filter",
    "Number of samples after bisulfiteConversion filter",
    "Number of samples after min. age 65 filter",
    "Number of samples before MALE PCA",
    "Number of samples after MALE PCA"
    
  )
)      
df.samples 
```


## Female
```{R, eval = TRUE}
df.samples <- data.frame(
  "Number of samples" =  c(
    nb.samples.female, 
    nb.samples.female, 
    nb.samples.min.age.65.female,
    nb.samples.before.pca.female, 
    nb.samples.after.pca.female
    
  ),
  "Description" = c(
    "Number of samples before bisulfiteConversion filter",
    "Number of samples after bisulfiteConversion filter",
    "Number of samples after min. age 65 filter",
    "Number of samples before FEMALE PCA",
    "Number of samples after FEMALE PCA"
    
  )
)    
df.samples 
```

## Number of probes removed in each step

```{R summary_table, eval = TRUE}
# Create summary table
df.probes <- data.frame(
  "Number of probes" = c(
    nb.probes.before.norm,
    length(passedProbes),
    nb.probes.after.norm
  ),
  "Description" = c(
    "total number of probes in raw data",
    "probes with detection P < 0.01 for all samples",
    "only probes that start with cg, no crosshyb, no smoking, mafcut = 0.01"
  )
)
df.probes %>% gt::gt()
```

# Checking the data

```{R checking}
aibl.se.with.sex <- get(load(
  file.path(dir.data.processed, "AIBL_se_age_at_least_65_with_XY.rda")
))
```

## show probes not starting with cg are removed
```{R}
grep("cg",rownames(aibl.se.with.sex),invert = T)
```

## show probes we have X/Y 
```{R}
c("chrX", "chrY") %in% (seqnames(aibl.se.with.sex) %>% as.character())
```

## show probes with SNP with MAF >= 0.01 and in the last 5bp were removed (compared with DMRcateData package probes)
```{R}
eh <- ExperimentHub()
snpsall <- eh[["EH3130"]]
any(snpsall$probe[snpsall$mafs > 0.01 & snpsall$distances <= 5] %>% as.character() %in% rownames(aibl.se.with.sex)) 
crosshyb <- eh[["EH3129"]]
any(crosshyb %in% rownames(aibl.se.with.sex))
XY.probes <- eh[["EH3131"]]
any(XY.probes %in% rownames(aibl.se.with.sex)) 
```

## select a few smoking probes, and show they are removed
```{R}
any(
  c("cg16145216", "cg19406367",
    "cg05603985", "cg10768063",
    "cg14099685", "cg12513616",
    "cg04085571", "cg23648810"
  ) %in% rownames(aibl.se.with.sex)
)
```

## show quantiles for each sample is the same
```{R}
boxplot(assay(aibl.se.with.sex[,1:10]))
plyr::adply(assay(aibl.se.with.sex[,1:10]),.margins = 2,.fun = function(x){quantile(x)})
```


## check cell type proportions adds up to 1
```{R}
table(rowSums(colData(aibl.se.with.sex)[,c("B","NK","CD4T","CD8T","Mono","Neutro","Eosino")] %>% as.matrix()))
```


## Check prediction sex

```{R}
aibl.prediction <- wateRmelon::estimateSex(assay(aibl.se.with.sex))
plyr::count(aibl.se.with.sex$Sex == gsub("emale|ale","",aibl.prediction$predicted_sex))
```
## show plate is a factor
```{R}
class(aibl.se.with.sex$PLATE)
```

## check outliers samples are not in datset anymore
```{R}
outliers <- readr::read_csv(file.path(dir.data.pca, "AIBL_PCs_usingBetas_age_at_least_65.csv"),col_types = readr::cols()) %>%
  dplyr::filter(outlier_PC1 == 1 | outlier_PC2 == 1) %>% dplyr::pull(1) 
head(outliers)
head(colnames(aibl.se.with.sex))
any(outliers %in% colnames(aibl.se.with.sex))
```


## for one sample, check distribution of type I and type II probes
```{R checking_part2, eval = TRUE}
library(sm)

annotType <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19::Manifest
annotType$designTypeNumeric <- ifelse(annotType$Type == "I",1,2)
probes <- assay(aibl.se.with.sex)[,1] %>% na.omit()
sm.density.compare(
  probes,
  annotType[names(probes),]$designTypeNumeric
)

probes <- assay(aibl.se.with.sex)[,2] %>% na.omit()
sm.density.compare(
  probes,
  annotType[names(probes),]$designTypeNumeric
)

```


## check the number of smoking probes that are included


```{R}
smoking.file <- "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5267325/bin/NIHMS817273-supplement-001506_-_Supplemental_Tables.xlsx"
if(!file.exists(file.path(dir.data.aux,basename(smoking.file)))) {
  downloader::download(smoking.file,file.path(dir.data.aux,basename(smoking.file)))
}
smoking <- readxl::read_xlsx(
  file.path(dir.data.aux,basename(smoking.file)),
  sheet = "02",
  skip = 2
)
smoking.sig.probes <- smoking %>%
  dplyr::filter(`P-value` < 1*10^(-7)) %>%
  dplyr::pull("Probe ID")

num.smoking.probes <- length(smoking.sig.probes)
num.smoking.probes.included <- length(rownames(aibl.se.with.sex)[rownames(aibl.se.with.sex) %in% smoking.sig.probes])
df.smoke <- data.frame(
  'Description' = c(
    "Total number of smoking probes",
    "Number of smoking probes included"),
  'Number of Sample' = c(
    num.smoking.probes,
    num.smoking.probes.included
  )
)    
df.smoke 
```

```{R checking_part5, include = FALSE}
rm(aibl.se.with.sex)
```

# Analysis

```{R}
aibl.se.with.sex <- get(load(
  file.path(dir.data.processed, "AIBL_se_age_at_least_65_with_XY.rda")
))
result_SSc <- readr::read_csv(file.path(dir.data,"AIBL_Smoking_Scores.csv"))
result_MSc <- readr::read_csv(file.path(dir.data,"AIBL_Mathylation_Scores.csv"))
McCartney_SSc <- readr::read_csv(file.path(dir.data,"AIBL_Smoking_Scores_McC.csv"))

aibl.se.with.sex <- aibl.se.with.sex[,aibl.se.with.sex$`disease status:ch1` %in% c("healthy control","Alzheimer's disease")]

pheno <- data.frame(colData (aibl.se.with.sex))
pheno$DIAGNOSIS <- factor(
  pheno$disease.status.ch1, 
  levels = c("healthy control", "Alzheimer's disease")
)
pheno$barcodes <- row.names(pheno)
pheno$Sex <- factor(pheno$Sex)
pheno$PLATE <- factor(pheno$PLATE)
pheno$granulocytes <- pheno$Eosino + pheno$Neutro
pheno <- pheno %>% dplyr::select(
  c(barcodes, Sex, age.pred.Elastic_Net, 
    DIAGNOSIS, PLATE,
    B, NK,   CD4T,   Mono, granulocytes
  )
)

pheno$smokingScore <- result_SSc[match(pheno$barcodes, result_SSc$SampleName),]$smokingScore
pheno$smokingScore2 <- result_MSc[match(pheno$barcodes, result_MSc$SampleName),]$methylationScore
pheno$smokingScore3 <- McCartney_SSc[match(pheno$barcodes, McCartney_SSc$SampleName),]$McC_SSc
```

## Metadata structure
```{R}
str(pheno)
```

## Number of samples
```{R}
plyr::count(pheno$DIAGNOSIS)
x <- plyr::count(pheno,c("Sex","PLATE"))
ggpubr::ggbarplot(data = x,x = "PLATE",y = "freq",facet.by = "Sex",color = "Sex",fill = "Sex") + ggtitle("AIBL")
plyr::count(pheno$DIAGNOSIS)
x <- plyr::count(pheno,c("Sex","DIAGNOSIS"))
ggpubr::ggbarplot(data = x,x = "DIAGNOSIS",y = "freq",facet.by = "Sex",color = "DIAGNOSIS",fill = "DIAGNOSIS") + ggtitle("AIBL") +  
  theme(
    axis.title.x = element_blank(),                                                                                                                        axis.text.x = element_blank(),
    axis.ticks.x = element_blank()
  )

```



```{R analysis_data, eval = TRUE}
# MALE
pheno.cn.ad.male <- pheno[pheno$DIAGNOSIS %in% c("healthy control", "Alzheimer's disease")  & pheno$Sex == "M",]
beta.percent.cn.ad.male <- assay(aibl.se.with.sex)[, pheno.cn.ad.male$barcodes] * 100

# FEMALE
pheno.cn.ad.female <- pheno[pheno$DIAGNOSIS %in% c("healthy control", "Alzheimer's disease") & pheno$Sex == "F",]
beta.percent.cn.ad.female <- assay(aibl.se.with.sex)[, pheno.cn.ad.female$barcodes] * 100

gc()
```


```{R}
plyr::count(pheno.cn.ad.female$DIAGNOSIS)
pheno.cn.ad.female %>% dplyr::group_by(DIAGNOSIS) %>% dplyr::summarise(mean(age.pred.Elastic_Net), sd(age.pred.Elastic_Net)) 
pheno.cn.ad.female %>% dplyr::summarise(mean(age.pred.Elastic_Net), sd(age.pred.Elastic_Net)) %>% round(digits = 1)
```

```{R}
plyr::count(pheno.cn.ad.male$DIAGNOSIS)
pheno.cn.ad.male %>% dplyr::group_by(DIAGNOSIS) %>% dplyr::summarise(mean(age.pred.Elastic_Net), sd(age.pred.Elastic_Net)) 
pheno.cn.ad.male %>% dplyr::summarise(mean(age.pred.Elastic_Net), sd(age.pred.Elastic_Net)) %>% round(digits = 1)
```

## Wilcoxon test on smoking scores for AD vs controls


### EpiSmokEr (Elliott et al. 2014)

```{r}
## Male
SSc.male.ad <- pheno.cn.ad.male %>% filter(DIAGNOSIS == "Alzheimer's disease") %>% dplyr::select(smokingScore)
SSc.male.cn <- pheno.cn.ad.male %>% filter(DIAGNOSIS == "healthy control") %>% dplyr::select(smokingScore)
wilcox.stats.male <- wilcox.test(SSc.male.ad$smokingScore, SSc.male.cn$smokingScore)
wilcox.stats.male
```


```{r}
## Female
SSc.female.ad <- pheno.cn.ad.female %>% filter(DIAGNOSIS == "Alzheimer's disease") %>% dplyr::select(smokingScore)
SSc.female.cn <- pheno.cn.ad.female %>% filter(DIAGNOSIS == "healthy control") %>% dplyr::select(smokingScore)
wilcox.stats.female <- wilcox.test(SSc.female.ad$smokingScore, SSc.female.cn$smokingScore)
wilcox.stats.female
```


```{r plot_SSc, echo = F}
ggplot(pheno, aes(x = DIAGNOSIS,y = smokingScore, fill = DIAGNOSIS)) + 
  geom_boxplot() + geom_jitter(shape=16, position=position_jitter(0.2)) +
  ggpubr::stat_compare_means(method = "wilcox.test") +
  theme_classic() + theme(legend.position="top") + 
  scale_fill_brewer(palette="Dark2") +
  facet_grid(~Sex) + ylab('Smoking Scores')
```


### EpiSmokEr (Zhang et al 2016)


```{r}
## Male
SSc.male.ad <- pheno.cn.ad.male %>% filter(DIAGNOSIS == "Alzheimer's disease") %>% dplyr::select(smokingScore2)
SSc.male.cn <- pheno.cn.ad.male %>% filter(DIAGNOSIS == "healthy control") %>% dplyr::select(smokingScore2)
wilcox.stats.male <- wilcox.test(SSc.male.ad$smokingScore2, SSc.male.cn$smokingScore2)
wilcox.stats.male
```


```{r}
## Female
SSc.female.ad <- pheno.cn.ad.female %>% filter(DIAGNOSIS == "Alzheimer's disease") %>% dplyr::select(smokingScore2)
SSc.female.cn <- pheno.cn.ad.female %>% filter(DIAGNOSIS == "healthy control") %>% dplyr::select(smokingScore2)
wilcox.stats.female <- wilcox.test(SSc.female.ad$smokingScore2, SSc.female.cn$smokingScore2)
wilcox.stats.female
```

```{r plot_SSc2, echo = F}
ggplot(pheno, aes(x = DIAGNOSIS,y = smokingScore2, fill = DIAGNOSIS)) + 
  geom_boxplot() + geom_jitter(shape=16, position=position_jitter(0.2)) +
  ggpubr::stat_compare_means(method = "wilcox.test") +
  theme_classic() + theme(legend.position="top") + 
  scale_fill_brewer(palette="Dark2") +
  facet_grid(~Sex) + ylab('Smoking Scores')
```


### Smoking scores (McCartney et al. 2018)

```{r}
## Male
SSc.male.ad <- pheno.cn.ad.male %>% filter(DIAGNOSIS == "Alzheimer's disease") %>% dplyr::select(smokingScore3)
SSc.male.cn <- pheno.cn.ad.male %>% filter(DIAGNOSIS == "healthy control") %>% dplyr::select(smokingScore3)
wilcox.stats.male <- wilcox.test(SSc.male.ad$smokingScore3, SSc.male.cn$smokingScore3)
wilcox.stats.male
```

```{r}
## Female
SSc.female.ad <- pheno.cn.ad.female %>% filter(DIAGNOSIS == "Alzheimer's disease") %>% dplyr::select(smokingScore3)
SSc.female.cn <- pheno.cn.ad.female %>% filter(DIAGNOSIS == "healthy control") %>% dplyr::select(smokingScore3)
wilcox.stats.female <- wilcox.test(SSc.female.ad$smokingScore3, SSc.female.cn$smokingScore3)
wilcox.stats.female
```


```{r plot_SSc3, echo = F}
ggplot(pheno, aes(x = DIAGNOSIS,y = smokingScore3, fill = DIAGNOSIS)) + 
  geom_boxplot() + geom_jitter(shape=16, position=position_jitter(0.2)) +
  ggpubr::stat_compare_means(method = "wilcox.test") +
  theme_classic() + theme(legend.position="top") + 
  scale_fill_brewer(palette="Dark2") +
  facet_grid(~Sex) + ylab('Smoking Scores')
```

## Models

### Logistic regression model - Using beta values

#### Male
```{R, eval = FALSE}
doParallel::registerDoParallel(cores = 6)
results.ad.glm.male <- plyr::adply(.data = beta.percent.cn.ad.male,.margins = 1, .fun = function(one){
  dat.cn.ad <- data.frame (beta = one, pheno.cn.ad.male)
  b <- glm(DIAGNOSIS ~ beta + age.pred.Elastic_Net + as.factor(PLATE) + B + NK + CD4T + Mono + granulocytes, data = dat.cn.ad, family = binomial)
  summary(b)
  beta_results <- coef(summary(b)) [grep("beta",rownames(coef(summary(b)) ),value = TRUE),]
  return(beta_results)
},.parallel = TRUE, .progress = "time",.id = "cpg",.inform = FALSE)

readr::write_csv(
  x = results.ad.glm.male, 
  file = file.path(dir.results.glm,"MALE_AIBL_glm_diagnosis_vs_beta_and_covaritates_age_at_least_65_AD_vs_CN_results.csv")
)
```

#### Female
```{R, eval = FALSE}

results.ad.glm.female <- plyr::adply(
  .data = beta.percent.cn.ad.female,.margins = 1, .fun = function(one){
    dat.cn.ad <- data.frame (beta = one, pheno.cn.ad.female)
    b <- glm(DIAGNOSIS ~ beta + age.pred.Elastic_Net + as.factor(PLATE) + B + NK + CD4T + Mono + granulocytes, data = dat.cn.ad, family = binomial)
    summary(b)
    beta_results <- coef(summary(b)) [grep("beta",rownames(coef(summary(b)) ),value = TRUE),]
    return(beta_results)
  },.parallel = TRUE, .progress = "time",.id = "cpg",.inform = FALSE)

readr::write_csv(
  x = results.ad.glm.female, 
  file = file.path(dir.results.glm,"FEMALE_AIBL_glm_diagnosis_vs_beta_and_covaritates_age_at_least_65_AD_vs_CN_results.csv")
)
```

## Annotate results

### Logistic regression model - Using beta values
```{R, eval = FALSE}
load(file.path(dir.data.aux,"great_EPIC_array_annotation.rda"))
load(file.path(dir.data.aux,"meta_analysis_cpgs.rda"))

annotate_results <- function(result, data.se){
  result$fdr <- p.adjust(result$`Pr(>|z|)`,method = "fdr")
  result <- cbind(
    result,
    rowRanges(data.se)[result$cpg] %>% as.data.frame() %>% dplyr::select(1:4)
  )
  result$Islands.UCSC.Relation_to_Island <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19::Islands.UCSC[result$cpg,"Relation_to_Island"]
  result$UCSC_RefGene_Name <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19::Other[result$cpg,"UCSC_RefGene_Name"]       
  result$UCSC_RefGene_Group <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19::Other[result$cpg,"UCSC_RefGene_Group"]     
  
  result <- dplyr::left_join(result, great,by = c("seqnames","start","end","cpg"))
  
  # sig.in.brain: 1 if it overlaps with the 3751 CpGs or 119 DMRs from our brain samples meta-analysis
  result$sig.in.brain <- result$cpg %in% brain.meta.analysis.cpgs
  return(result)
}

aibl.se <- readRDS(
  file.path(dir.data.pca, "AIBL_QNBMIQ_PCfiltered_age_at_least_65.RDS")
)

results.ad.glm.female <- readr::read_csv(
  file = file.path(
    dir.results.glm,
    "FEMALE_AIBL_glm_diagnosis_vs_beta_and_covaritates_age_at_least_65_AD_vs_CN_results.csv")
)
colnames(results.ad.glm.female)[1] <- "cpg"
results.ad.glm.female.annotated <- results.ad.glm.female %>% annotate_results(aibl.se.with.sex)
readr::write_csv(
  x = results.ad.glm.female.annotated,
  file = file.path(dir.results.glm,"FEMALE_AIBL_glm_diagnosis_vs_beta_and_covaritates_age_at_least_65_AD_vs_CN_results_annotated.csv")
)


results.ad.glm.male <- readr::read_csv(
  file = file.path(
    dir.results.glm,
    "MALE_AIBL_glm_diagnosis_vs_beta_and_covaritates_age_at_least_65_AD_vs_CN_results.csv")
)
colnames(results.ad.glm.male)[1] <- "cpg"
results.ad.glm.male.annotated <- results.ad.glm.male %>% annotate_results(aibl.se.with.sex)

readr::write_csv(
  x = results.ad.glm.male.annotated,
  file = file.path(dir.results.glm,"MALE_AIBL_glm_diagnosis_vs_beta_and_covaritates_age_at_least_65_AD_vs_CN_results_annotated.csv")
)

results.ad.glm.male.smok <- readr::read_csv(
  file = file.path(
    dir.results.glm,
    "MALE_AIBL_glm_diagnosis_vs_beta_and_smoking_covaritates_age_at_least_65_AD_vs_CN_results.csv")
)
colnames(results.ad.glm.male.smok)[1] <- "cpg"
results.ad.glm.male.smok.annotated <- results.ad.glm.male.smok %>% annotate_results(aibl.se.with.sex)

readr::write_csv(
  x = results.ad.glm.male.smok.annotated,
  file = file.path(dir.results.glm,"MALE_AIBL_glm_diagnosis_vs_beta_smoking_and_covaritates_age_at_least_65_AD_vs_CN_results_annotated.csv")
)

results.ad.glm.female.smok <- readr::read_csv(
  file = file.path(
    dir.results.glm,
    "FEMALE_AIBL_glm_diagnosis_vs_beta_smoking_and_covaritates_age_at_least_65_AD_vs_CN_results.csv")
)
colnames(results.ad.glm.female.smok)[1] <- "cpg"
results.ad.glm.female.smok.annotated <- results.ad.glm.female.smok %>% annotate_results(aibl.se.with.sex)

readr::write_csv(
  x = results.ad.glm.female.smok.annotated,
  file = file.path(dir.results.glm,"FEMALE_AIBL_glm_diagnosis_vs_beta_smoking_and_covaritates_age_at_least_65_AD_vs_CN_results_annotated.csv")
)
```


# Calculate inflation factors and Bacon correction for final results

```{R read_results, include = FALSE}
results.ad.glm.female.annotated <- readr::read_csv(
  file.path(dir.results.glm,"FEMALE_AIBL_glm_diagnosis_vs_beta_and_covaritates_age_at_least_65_AD_vs_CN_results_annotated.csv")
)

results.ad.glm.male.annotated <- readr::read_csv(
  file.path(dir.results.glm,"MALE_AIBL_glm_diagnosis_vs_beta_and_covaritates_age_at_least_65_AD_vs_CN_results_annotated.csv")
)

results.ad.glm.female.smok.annotated <- readr::read_csv(
  file.path(dir.results.glm,"FEMALE_AIBL_glm_diagnosis_vs_beta_smoking_and_covaritates_age_at_least_65_AD_vs_CN_results_annotated.csv")
)

results.ad.glm.male.smok.annotated <- readr::read_csv(
  file.path(dir.results.glm,"MALE_AIBL_glm_diagnosis_vs_beta_smoking_and_covaritates_age_at_least_65_AD_vs_CN_results_annotated.csv")
)
```

```{R, message = FALSE, warning = FALSE, result = "hide"}
library(dplyr)
library(bacon)
library(GWASTools)
```

## Auxiliary function

```{R inflation}
estimation_of_inflation <- function(data){
  ### 1. Compute genomic inflation factor before bacon adjustment
  data$zvalue <- data$Estimate / data$`Std. Error`
  #data$zvalue <- qnorm(pt(data$tvalue, df))
  data$chisq <- (data$zvalue) ^ 2
  # inflation factor - last term is median from chisq distrn with 1 df  
  inflationFactor <- median(data$chisq,na.rm = TRUE) / qchisq(0.5, 1)
  print("o Before bacon correction")
  print("Conventional lambda")
  print(inflationFactor)
  
  bc <- bacon(
    teststatistics = NULL,
    effectsizes =  data$Estimate,
    standarderrors = data$`Std. Error`,
    na.exclude = TRUE
  )
  # inflation factor
  print("bacon lambda")
  print(inflation(bc))
  print(estimates(bc))
  
  ### 3. Create final dataset
  data.with.inflation <- data.frame(
    data,
    Estimate.bacon = bacon::es(bc),
    StdErr.bacon = bacon::se(bc),
    pValue.bacon = pval(bc),
    fdr.bacon = p.adjust(pval(bc), method = "fdr"),
    stringsAsFactors = FALSE
  )
  print("o After bacon correction")
  print("Conventional lambda")
  print(median((data.with.inflation$Estimate.bacon/data.with.inflation$StdErr.bacon) ^ 2,na.rm = TRUE)/qchisq(0.5, 1))
  
  percent_null <- trunc ( estimates(bc)[1]*100, digits = 0)
  percent_1  <- trunc ( estimates(bc)[2]*100, digits = 0 )
  percent_2  <- 100 - percent_null - percent_1  
  bc2 <- bacon(
    teststatistics = NULL,
    effectsizes =  data.with.inflation$Estimate.bacon,
    standarderrors = data.with.inflation$StdErr.bacon,
    na.exclude = TRUE, 
    priors = list(
      sigma = list(alpha = 1.28,  beta = 0.36), 
      mu = list(lambda = c(0, 3, -3), tau = c(1000, 100, 100)), 
      epsilon = list(gamma = c(percent_null, percent_1, percent_2)))
  )
  print("inflation")
  print(inflation(bc2))
  print("estimates")
  print(estimates(bc2))
  
  
  data.with.inflation <- data.with.inflation %>% dplyr::select(-c(zvalue, chisq))
  return(
    list(
      "data.with.inflation" = data.with.inflation,
      "inflationFactor" =   inflationFactor,
      "estimatedInflation" = inflation(bc)
    )
  )
}


plot_inflation <- function(data, main){
  qqPlot(
    data$data.with.inflation$Pr...t.., ci = TRUE, 
    main = paste0(main)
  ) + text(
    x = 0.5, 
    y = -log10(min(data$data.with.inflation$Pr...t..)) - 0.5,
    labels = bquote(lambda == .(format(data$inflationFactor,digits = 4))),
    pos = 4
  ) + text(
    x = 0.5, 
    y = -log10(min(data$data.with.inflation$Pr...t..)) - 1,
    labels = bquote(lambda[bacon] == .(format(data$estimatedInflation,digits = 4))),
    pos = 4
  )
}
```

## Logistic regression model - No smoking scores

### Female
```{R,fig.width = 7}
results.ad.glm.female.with.inflation <- estimation_of_inflation(data = results.ad.glm.female.annotated)
```

```{r eval = F}
readr::write_csv(
  x = results.ad.glm.female.with.inflation$data.with.inflation,
  file = file.path(dir.results.glm,"FEMALE_AIBL_glm_diagnosis_vs_beta_and_covaritates_age_at_least_65_AD_vs_CN_results_annotated_with_inflation.csv")
)

#plot_inflation(
#  data = results.ad.glm.female.with.inflation, 
#  main = "FEMALE AIBL AD vs CN - single cpg\n Logistic regression model"
#)
```

```{r}
qqPlot(
  results.ad.glm.female.with.inflation$data.with.inflation$pValue.bacon, ci = TRUE, 
  main = paste0("FEMALE - AIBL AD vs CN - single cpg\nlogistic regression model - pValue.bacon")
) 
```

### Male
```{R,fig.width = 7}
results.ad.glm.male.with.inflation <- estimation_of_inflation(data = results.ad.glm.male.annotated)
```

```{r eval = F}
readr::write_csv(
  x = results.ad.glm.male.with.inflation$data.with.inflation,
  file = file.path(dir.results.glm,"MALE_AIBL_glm_diagnosis_vs_beta_and_covaritates_age_at_least_65_AD_vs_CN_results_annotated_with_inflation.csv")
)
```


```{r}
qqPlot(
  results.ad.glm.male.with.inflation$data.with.inflation$pValue.bacon, ci = TRUE, 
  main = paste0("MALE - AIBL AD vs CN - single cpg\nlogistic regression model - pValue.bacon")
) 
```

## Logistic regression model - Smoking scores

### Female
```{R,fig.width = 7}
results.ad.glm.female.smok.with.inflation <- estimation_of_inflation(data = results.ad.glm.female.smok.annotated)
```

```{r eval = F}
readr::write_csv(
  x = results.ad.glm.female.smok.with.inflation$data.with.inflation,
  file = file.path(dir.results.glm,"FEMALE_AIBL_glm_diagnosis_vs_beta_smoking_and_covaritates_age_at_least_65_AD_vs_CN_results_annotated_with_inflation.csv")
)

#plot_inflation(
#  data = results.ad.glm.female.with.inflation, 
#  main = "FEMALE AIBL AD vs CN - single cpg\n Logistic regression model"
#)
```

```{r}
qqPlot(
  results.ad.glm.female.smok.with.inflation$data.with.inflation$pValue.bacon, ci = TRUE, 
  main = paste0("FEMALE - AIBL AD vs CN - single cpg\nlogistic regression model - pValue.bacon")
) 
```

### Male
```{R,fig.width = 7}
results.ad.glm.male.smok.with.inflation <- estimation_of_inflation(data = results.ad.glm.male.smok.annotated)
```

```{r eval = F}
readr::write_csv(
  x = results.ad.glm.male.smok.with.inflation$data.with.inflation,
  file = file.path(dir.results.glm,"MALE_AIBL_glm_diagnosis_vs_beta_smoking_and_covaritates_age_at_least_65_AD_vs_CN_results_annotated_with_inflation.csv")
)
```


```{r}
qqPlot(
  results.ad.glm.male.smok.with.inflation$data.with.inflation$pValue.bacon, ci = TRUE, 
  main = paste0("MALE - AIBL AD vs CN - single cpg\nlogistic regression model - pValue.bacon")
) 
```

# Session information
```{R}
devtools::session_info()
```