---
title: "Integrative meta-analysis of epigenome-wide association studies identifies genomic and epigenomics differences in the brain and the blood in Alzheimerâ€™s disease - Stratified by Sex"
subtitle: 'ADNI dataset'
author:
  - Tiago Chedraoui Silva^[University of Miami]
  - Lily Wang^[University of Miami]
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  rmarkdown::html_document:
    highlight: breezedark
    theme: lumen
    toc: true
    number_sections: true
    df_print: paged
    code_download: false
    toc_float:
      collapsed: yes
    toc_depth: 3
editor_options:
  chunk_output_type: inline    
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,warning = FALSE)
knitr::opts_knit$set(root.dir = "./../../")
```

# Setup

```{R Setup, message = FALSE, results = 'hide'}
library(dplyr)
library(SummarizedExperiment)
library(ExperimentHub)
library(ADNIMERGE)
library(dplyr)
library(readxl)
library(wateRmelon)
library(minfi)
cohort <- "ADNI"
dir.base <- getwd()
dir.base.data <- file.path(dir.base,"../AD-meta-analysis-blood-samples/")
dir.data <- file.path(dir.base.data,"datasets/",cohort,"/data/DNA_methylation") 
dir.data.clinical <- file.path(dir.base.data,"datasets/",cohort,"/data/Clinical") 
dir.data.raw <- file.path(dir.base.data,"datasets/",cohort,"/data/DNA_methylation/idat") 
dir.data.processed <- file.path(dir.base.data,"datasets/",cohort,"/data/DNA_methylation/processed") 
dir.data.aux <- file.path(dir.base.data,"datasets/Aux/") 
dir.data.pca <- file.path(dir.data,"/pca_filtering/") 
dir.results <- file.path(dir.base,"analysis_results",cohort)
dir.results.glm <- file.path(dir.results,"/Logistic_regression_model")
for(p in grep("dir",ls(),value = T)) dir.create(get(p),recursive = TRUE,showWarnings = FALSE)
```

# Data retrieval

```{R, load_rgset_bs_filtered_min_65, include = FALSE, eval = FALSE}
load(file.path(dir.data.processed,"RGSet_bs_filtered_min_age_at_visit_65.rda"))
```

```{R, load_bs_filtered_min_65, include = FALSE, eval = FALSE}
adni.se <- readRDS(file.path(dir.data.pca, "ADNI_QNBMIQ_PCfiltered_min_age_at_visit_65.RDS"))
```


```{R, load_detectP, include = FALSE, eval = FALSE}
load(file.path(dir.data.processed,"detectionP_passed_probes.rda"))
nb.probes.before.norm <- nrow(failed.01)
# length(passedProbes)
# length(grep("cg",passedProbes))
```

## Probes QC

1. keep only probes that start with "cg"
2. drop probes that are on X/Y
3. drop probes where SNP with MAF >= 0.01 was present in the last 5 bp of the probe.
4. Remove smoking related probes

```{R, eval = FALSE}
library(minfi)
library(wateRmelon)
library(RPMM)
library(DMRcate)
```

```{R, eval = FALSE}
annotation <- minfi::getAnnotation(RGSet.bs.filtered.min.age_at_visit.65) %>% as.data.frame %>% dplyr::filter(chr %in% c("chrX","chrY"))
betaSet <- getBeta(RGSet.bs.filtered.min.age_at_visit.65)
betaSet <- betaSet[rownames(betaSet) %in% rownames(annotation), ]
nb.chrXY.probes <- nrow(betaSet)
rm(RGSet.bs.filtered.min.age_at_visit.65)
betaSet <- betaSet[rownames(betaSet) %in% passedProbes, ]
betaSet <- betaSet[grep("cg",rownames(betaSet)),]
gc()

# snapshotDate(): 2021-03-30
betaSet <- rmSNPandCH(
  object = betaSet,
  dist = 5,
  mafcut = 0.01,
  and = TRUE,
  rmcrosshyb = TRUE,
  rmXY = FALSE
)
```

```{R, include = F}
# > dim(betaSet)
# [1] 15782  1516
```

## Remove smoking probes

```{R, eval = FALSE}
smoking.file <- "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5267325/bin/NIHMS817273-supplement-001506_-_Supplemental_Tables.xlsx"
if(!file.exists(file.path(dir.data.aux,basename(smoking.file)))) {
  downloader::download(smoking.file,file.path(dir.data.aux,basename(smoking.file)))
}
smoking <- readxl::read_xlsx(
  file.path(dir.data.aux,basename(smoking.file)),
  sheet = "02",
  skip = 2
)
smoking.sig.probes <- smoking %>%
  dplyr::filter(`P-value` < 1*10^(-7)) %>%
  dplyr::pull("Probe ID")

betaSet <- betaSet[!rownames(betaSet) %in% smoking.sig.probes,]
```

```{R, include = F}
# > dim(betaSet)
# [1] 15777  1516
```


## Merge DNA methylation matrix on autosome with the one on chrX, chrY

```{R, eval = FALSE}
betaSet <- betaSet[,colnames(adni.se)]
identical(colnames(adni.se), colnames(betaSet))
betaQN_BMIQ <- rbind(assay(adni.se), betaSet)
table(base::duplicated(row.names(betaQN_BMIQ)))
# 744109
save(
  betaQN_BMIQ,
  clinical.filtered.min.age_at_visit.65,
  file = file.path(dir.data.processed,"RGSet_bs_filtered_min_age_at_visit_65_with_XY.rda")
)
```

### Summarize Experiment

```{R, eval = FALSE}
library(ADNIMERGE)
library(dplyr)
library(readxl)
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
library(S4Vectors)
epic.hg19 <- minfi::getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
epic.hg19.gr <- epic.hg19 %>% makeGRangesFromDataFrame(
  start.field = "pos", end.field = "pos",keep.extra.columns = T
)
clinical <- colData(adni.se)
betaQN_BMIQ <- betaQN_BMIQ[rownames(betaQN_BMIQ) %in% names(epic.hg19.gr),]
rowData <- epic.hg19.gr[rownames(betaQN_BMIQ),]
data <- betaQN_BMIQ[,colnames(betaQN_BMIQ) %in% rownames(clinical)]
data <- data[,rownames(clinical)]

adni.se <- SummarizedExperiment::SummarizedExperiment(
  assays = SimpleList("DNAm" = data),
  rowData = rowData,
  metadata = list("genome" = "hg19"),
  colData = clinical
)

saveRDS(adni.se, file = file.path(dir.data.processed,"adni_se_min_age_at_visit_65_with_XY.RDS"))
readr::write_csv(
  clinical %>% as.data.frame, 
  file = file.path(dir.data.processed,"ADNI_Se_colData_with_XY.csv")
)
```


# Summary after QC steps


```{R, eval = FALSE}
adni.se <- readRDS(file.path(dir.data.processed,"adni_se_min_age_at_visit_65_with_XY.RDS"))
adni.se <- adni.se[,adni.se$DX %in% c("CN","Dementia")]
pheno_df <- colData(adni.se) %>% as.data.frame
nb.samples.before.pca.male <- sum(pheno_df$PTGENDER == "Female")
nb.samples.before.pca.female <- sum(pheno_df$PTGENDER == "Male")
```

## Data and metadata
```{R}
adni.se <- readRDS(file.path(dir.data.processed,"adni_se_min_age_at_visit_65_with_XY.RDS"))
adni.se <- adni.se[,adni.se$DX %in% c("CN","Dementia")]
nb.samples.after.pca <- ncol(adni.se)
nb.probes.after.norm <- nrow(adni.se)
```



```{R}
pheno.cn.ad <- colData(adni.se)
pheno.cn.ad.female <- pheno.cn.ad[pheno.cn.ad$PTGENDER == "Female" & pheno.cn.ad$DX %in% c("CN","Dementia"),]
nb.samples.after.pca.female <- nrow(pheno.cn.ad.female)
plyr::count(pheno.cn.ad.female$DX)
pheno.cn.ad.female %>% as.data.frame %>% dplyr::group_by(DX) %>% dplyr::summarise(mean(age_at_visit), sd(age_at_visit))
pheno.cn.ad.female %>% as.data.frame %>% dplyr::summarise(mean(age_at_visit), sd(age_at_visit))  %>% round(digits = 1)
```

```{R}
pheno.cn.ad.male <- pheno.cn.ad[pheno.cn.ad$PTGENDER == "Male" & pheno.cn.ad$DX %in% c("CN","Dementia"),]
plyr::count(pheno.cn.ad.male$DX)
nb.samples.after.pca.male <- nrow(pheno.cn.ad.male)
pheno.cn.ad.male %>% as.data.frame %>% dplyr::group_by(DX) %>% dplyr::summarise(mean(age_at_visit), sd(age_at_visit))
pheno.cn.ad.male %>% as.data.frame %>% dplyr::summarise(mean(age_at_visit), sd(age_at_visit))  %>% round(digits = 1)
```


## Numbers of samples and probes removed in each step

### Female
```{R, eval = FALSE, echo=FALSE, include = F}
df.samples <- data.frame(
  "Number of samples" =  c(
    nb.samples, 
    nb.samples.bc.filtered, 
    nb.samples.before.pca, 
    nb.samples.after.pca.female
  ),
  "Description" = c(
    "Number of FEMALE samples before bisulfiteConversion filter",
    "Number of FEMALE samples after bisulfiteConversion filter",
    "Number of FEMALE samples before PCA",
    "Number of FEMALE samples after PCA"
  )
)    
df.samples 
```

### Male
```{R, eval = FALSE, echo=FALSE, include = F}
df.samples <- data.frame(
  "Number of samples" =  c(
    nb.samples, 
    nb.samples.bc.filtered, 
    nb.samples.before.pca, 
    nb.samples.after.pca.male
  ),
  "Description" = c(
    "Number of MALE samples before bisulfiteConversion filter",
    "Number of MALE samples after bisulfiteConversion filter",
    "Number of MALE samples before PCA",
    "Number of MALE samples after PCA"
  )
)    
df.samples 
```


```{R, eval = FALSE, echo=FALSE, include = F}
# Create summary table
df.probes <- data.frame(
  "Number of probes" = c(
    nb.probes.before.norm,
    length(passedProbes),
    nb.probes.after.norm
  ),
  "Description" = c(
    "total number of probes in raw data",
    "probes with detection P < 0.01 for all samples",
    "only probes that start with cg, no crosshyb, no X, no Y, no smoking, mafcut = 0.01"
  )
)
df.probes %>% gt::gt() 
```

# Checking the data

## show probes not starting with cg are removed
```{R check_cg}
grep("cg",rownames(adni.se),invert = T)
```

## show probes on X/Y are removed
```{R check_xy}
c("chrX", "chrY") %in% (seqnames(adni.se) %>% as.character())
```

## show probes with SNP with MAF >= 0.01 and in the last 5bp were removed (compared with DMRcateData package probes)
```{R}
eh <- ExperimentHub()
snpsall <- eh[["EH3130"]]
any(snpsall$probe[snpsall$mafs > 0.01 & snpsall$distances <= 5] %>% as.character() %in% rownames(adni.se)) 
crosshyb <- eh[["EH3129"]]
any(crosshyb %in% rownames(adni.se))
XY.probes <- eh[["EH3131"]]
any(XY.probes %in% rownames(adni.se)) 
```

## select a few smoking probes, and show they are removed
```{R}
any(
  c("cg16145216", "cg19406367",
    "cg05603985", "cg10768063",
    "cg14099685", "cg12513616",
    "cg04085571", "cg23648810"
  ) %in% rownames(adni.se)
)
```

## show quantiles for each sample is the same
```{R quantile_box_plot}
boxplot(assay(adni.se[,1:10]))
plyr::adply(assay(adni.se[,1:10]),.margins = 2,.fun = function(x){quantile(x)})
```


## check cell type proportions adds up to 1
```{R}
table(rowSums(colData(adni.se)[,c("B","NK","CD4T","CD8T","Mono","Neutro","Eosino")] %>% as.matrix()))
```

## show plate is a factor
```{R}
class(adni.se$PlateNumber)
```

## check outliers samples are not in datset anymore
```{R}
outliers <- readr::read_csv(paste0(dir.data.pca, "ADNI_PCs_usingBetas_min_age_at_visit_65.csv"),col_types = readr::cols()) %>%
  dplyr::filter(outlier_PC1 == 1 | outlier_PC2 == 1) %>% pull(1) 
head(outliers)
head(colnames(adni.se))
any(outliers %in% colnames(adni.se))
```

## For one sample, check distribution of type I and type II probes
```{R checking_part2, eval = TRUE}
library(sm)

annotType <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19::Manifest
annotType$designTypeNumeric <- ifelse(annotType$Type == "I",1,2)
probes <- assay(adni.se)[,1] %>% na.omit()
sm.density.compare(
  probes,
  annotType[names(probes),]$designTypeNumeric
)

probes <- assay(adni.se)[,2] %>% na.omit()
sm.density.compare(
  probes,
  annotType[names(probes),]$designTypeNumeric
)

```


```{R checking_part5, include = FALSE}
rm(adni.se)
```


# Analysis
```{R load_analysis_data, eval = FALSE}
adni.se <- readRDS(file.path(dir.data.processed,"adni_se_min_age_at_visit_65_with_XY.RDS"))
```


```{R sas, eval = FALSE}
str_data <- function(se){
  
  mvalues <- log2(assay(se)/(1 - assay(se)))
  mvalues <- mvalues %>% t %>% as.data.frame()
  mvalues$barcodes <- rownames(mvalues)
  mvalues.pivot <- pivot_longer(mvalues,!barcodes,names_to = "cpg", values_to = "mvalues")
  
  beta <- assay(se)
  beta <- beta %>% t %>% as.data.frame()
  beta$barcodes <- rownames(beta)
  beta.pivot <- pivot_longer(beta,!barcodes,names_to = "cpg", values_to = "beta")
  
  
  #  0 = control, 1 = AD, 2 = MCI
  pheno <- data.frame(colData (se))
  pheno$DIAGNOSIS <- factor(pheno$DX, levels = c("CN", "MCI", "Dementia"))
  pheno <- pheno[!is.na(pheno$DIAGNOSIS),]
  pheno$barcodes <- row.names(pheno)
  
  # add indicator for first and last visit
  pheno$visit <- sub("m",  "", pheno$VISCODE)
  pheno$PlateNumber <- factor(pheno$PlateNumber)
  pheno$granulocytes <- pheno$Eosino + pheno$Neutro
  
  pheno <- pheno %>% dplyr::select(
    c(barcodes, RID, VISCODE, PTGENDER, age_at_visit, visit, 
      DIAGNOSIS, PlateNumber,
      B, NK,   CD4T, Mono,   granulocytes
    )
  )
  final <- dplyr::inner_join(pheno,beta.pivot) %>%  dplyr::inner_join(mvalues.pivot)
  final <- final[order(final$cpg,final$RID,final$VISCODE),]
  final
}

dir.create(file.path(dir.results,"AD_CN_male/DATA_for_SAS_models//"),recursive = T)
for(start in seq(1,nrow(adni.se),by = 1000)){
  print(start)
  end <- start + 1000 - 1
  idx <- which(start == seq(1,nrow(adni.se),by = 1000))
  if(end > nrow(adni.se)) end <- nrow(adni.se)
  adni.se.selected <- adni.se[start:end, adni.se$DX %in% c("Dementia","CN") & adni.se$PTGENDER == "Male"]
  print(dim(adni.se.selected))
  final <- str_data(adni.se.selected)
  final$VISCODE <- final$visit <- final$mvalues <- NULL
  readr::write_csv(
    final,
    file = paste0(
      dir.results,
      "/AD_CN_male/DATA_for_SAS_models//",
      "ADNI_1000_cpgs_",idx,".csv"
    )
  )
}

dir.create(file.path(dir.results,"AD_CN_female/DATA_for_SAS_models//"),recursive = T)
plyr::a_ply(seq(1,nrow(adni.se),by = 1000),.margins = 1,.fun = function(start){
  end <- start + 1000 - 1
  idx <- which(start == seq(1,nrow(adni.se),by = 1000))
  if(end > nrow(adni.se)) end <- nrow(adni.se)
  adni.se.selected <- adni.se[start:end,adni.se$DX %in% c("Dementia","CN") & adni.se$PTGENDER == "Female"]
  final <- str_data(adni.se.selected)
  final$VISCODE <- final$visit <- final$mvalues <- NULL
  readr::write_csv(
    final,
    file = paste0(
      dir.results,
      "/AD_CN_female/DATA_for_SAS_models//",
      "ADNI_1000_cpgs_",idx,".csv"
    )
  )
  
},.progress = "time")



```


## Logistic regression model


### AD vs CN 

#### Male
```{R, eval = FALSE}
files <- dir(
  file.path(dir.base,"LW/Males/"),
  full.names = T,
  recursive = T,
  pattern = "_[0-9]*.csv"
)
length(files)
glmm <- plyr::adply(
  .data = files,
  .margins = 1,
  .fun = function(f) {
    readr::read_csv(f,col_types = readr::cols())
  },.progress = "time",.id = NULL) %>% unique()  
readr::write_csv(
  glmm, 
  file = file.path(dir.results.glm,"Males_ADNI_AD_vs_CN_glmm_using_beta_SAS.cvs")
)
```

#### Female
```{R, eval = FALSE}
files <- dir(
  file.path(dir.base,"LW/Females/"),
  full.names = T,
  recursive = T,
  pattern = "_[0-9]*.csv"
)
length(files)
glmm <- plyr::adply(
  .data = files,
  .margins = 1,
  .fun = function(f) {
    readr::read_csv(f,col_types = readr::cols())
  },.progress = "time",.id = NULL) %>% unique()  
readr::write_csv(
  glmm, 
  file = file.path(dir.results.glm,"Females_ADNI_AD_vs_CN_glmm_using_beta_SAS.cvs")
)
```

## Annotate results
```{R annotate_functions, eval = FALSE}
message("o Stating to annotate")
adni.se <- readRDS(file.path(dir.data.processed, "adni_se_min_age_at_visit_65_with_XY.RDS"))
load(file.path(dir.data.aux,"great_EPIC_array_annotation.rda"))
load(file.path(dir.data.aux,"E073_15_coreMarks_segments.rda"))
load(file.path(dir.data.aux,"meta_analysis_cpgs.rda"))

annotate_results <- function(result, data.se){
  #result <- result[result$cpg %in% rownames(data.se),]
  result$fdr <- p.adjust(result$Probt,method = "fdr")
  result <- cbind(
    result,
    rowRanges(data.se)[result$cpg] %>% as.data.frame() %>% dplyr::select(1:4)
  )
  result$Islands.UCSC.Relation_to_Island <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19::Islands.UCSC[result$cpg,"Relation_to_Island"]
  result$UCSC_RefGene_Name <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19::Other[result$cpg,"UCSC_RefGene_Name"]       
  result$UCSC_RefGene_Group <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19::Other[result$cpg,"UCSC_RefGene_Group"]     
  
  result <- dplyr::left_join(result, great,by = c("seqnames","start","end","cpg"))
  
  hits <- findOverlaps(rowRanges(data.se)[result$cpg],ChmmModels.gr) %>% as.data.frame()
  hits$state <- ChmmModels.gr$state[hits$subjectHits]
  hits$cpg <- result$cpg[hits$queryHits]
  result <- merge(result, hits[,c("state","cpg")], all.x = TRUE, sort = FALSE, by = "cpg")
  
  result$sig.in.brain <- result$cpg %in% brain.meta.analysis.cpgs
  return(result)
}
```

```{R annotate, eval = FALSE}
logistic_mixed_model_ad_cn.females <- readr::read_csv( 
  file.path(dir.results.glm,"Females_ADNI_AD_vs_CN_glmm_using_beta_SAS.cvs")
)

logistic_mixed_model_ad_cn_sex <- readxl::read_xlsx( 
  file.path(dir.base,"LW/Females/results/ad-vs-cn-females-sex-chr_8-9-2021.xlsx")
) %>% dplyr::mutate(file_order = NULL)
logistic_mixed_model_ad_cn.females <- rbind(logistic_mixed_model_ad_cn.females, logistic_mixed_model_ad_cn_sex)


logistic_mixed_model_ad_cn.females <- logistic_mixed_model_ad_cn.females %>% annotate_results(adni.se)
readr::write_csv(
  x = logistic_mixed_model_ad_cn.females,
  file = file.path(dir.results.glm,"Females_ADNI_AD_vs_CN_glmm_using_beta_all_annotated.cvs"),
)

logistic_mixed_model_ad_cn.males <- readr::read_csv( 
  file.path(dir.results.glm,"Males_ADNI_AD_vs_CN_glmm_using_beta_SAS.cvs")
)

logistic_mixed_model_ad_cn_sex <- readxl::read_xlsx( 
  file.path(dir.base,"LW/Males/results/ad-vs-cn-males-sex-chr_8-9-2021.xlsx")
) %>% dplyr::mutate(file_order = NULL)

logistic_mixed_model_ad_cn.males <- rbind(logistic_mixed_model_ad_cn.males, logistic_mixed_model_ad_cn_sex)
logistic_mixed_model_ad_cn.annotated.males <- logistic_mixed_model_ad_cn.males %>% annotate_results(adni.se)
readr::write_csv(
  x = logistic_mixed_model_ad_cn.annotated.males,
  file = file.path(dir.results.glm,"Males_ADNI_AD_vs_CN_glmm_using_beta_all_annotated.cvs"),
)

```

```{R read_results, include = FALSE}
logistic_mixed_model_ad_cn.annotated.females <- readr::read_csv(
  file.path(dir.results.glm,"Females_ADNI_AD_vs_CN_glmm_using_beta_all_annotated.cvs")
)
logistic_mixed_model_ad_cn.annotated.males <- readr::read_csv(
  file.path(dir.results.glm,"Males_ADNI_AD_vs_CN_glmm_using_beta_all_annotated.cvs")
)
```

# Calculate inflation factors and Bacon correction for final results

## Auxiliary functions

```{R inflation, message = FALSE, results = 'hide'}
library(dplyr)
library(bacon)
library(GWASTools)

estimation_of_inflation <- function(data){
  ### 1. Compute genomic inflation factor before bacon adjustment
  data$zvalue <- data$Estimate / data$StdErr
  #data$zvalue <- qnorm(pt(data$tvalue, df))
  data$chisq <- (data$zvalue) ^ 2
  # inflation factor - last term is median from chisq distrn with 1 df  
  inflationFactor <- median(data$chisq,na.rm = TRUE) / qchisq(0.5, 1)
  print("o Before bacon correction")
  print("Conventional lambda")
  print(inflationFactor)
  ### 2. bacon analysis
  bc <- bacon(
    teststatistics = NULL,
    effectsizes =  data$Estimate,
    standarderrors = data$StdErr,
    na.exclude = TRUE
  )
  # inflation factor
  print("bacon lambda")
  print(inflation(bc))
  
  ### 3. Create final dataset
  data.with.inflation <- data.frame(
    data,
    Estimate.bacon = bacon::es(bc),
    StdErr.bacon = bacon::se(bc),
    pValue.bacon = pval(bc),
    fdr.bacon = p.adjust(pval(bc), method = "fdr"),
    stringsAsFactors = FALSE
  )
  print("o After bacon correction")
  print("Conventional lambda")
  print(median((data.with.inflation$Estimate.bacon/data.with.inflation$StdErr.bacon) ^ 2,na.rm = TRUE)/qchisq(0.5, 1))
  print("bacon lambda")
  
  percent_null <- trunc ( estimates(bc)[1]*100, digits = 0)
  percent_1  <- trunc ( estimates(bc)[2]*100, digits = 0 )
  percent_2  <- 100 - percent_null - percent_1  
  bc2 <- bacon(
    teststatistics = NULL,
    effectsizes =  data.with.inflation$Estimate.bacon,
    standarderrors = data.with.inflation$StdErr.bacon,
    na.exclude = TRUE, 
    priors = list(sigma = list(alpha = 1.28, beta =
                                 0.36), mu = list(lambda = c(0, 3, -3), tau = c(1000, 100, 100)), epsilon
                  = list(gamma = c(percent_null, percent_1, percent_2)))
  )
  
  print(inflation(bc2))
  print(estimates(bc2))
  
  data.with.inflation <- data.with.inflation %>% dplyr::select(-c(zvalue, chisq))
  return(
    list(
      "data.with.inflation" = data.with.inflation,
      "inflationFactor" =   inflationFactor,
      "estimatedInflation" = inflation(bc)
    )
  )
}


plot_inflation <- function(data, main){
  qqPlot(
    data$data.with.inflation$Probt, ci = TRUE, 
    main = paste0(main)
  ) + text(
    x = 0.5, 
    y = -log10(min(data$data.with.inflation$Probt)) - 0.5,
    labels = bquote(lambda == .(format(data$inflationFactor,digits = 4))),
    pos = 4
  ) + text(
    x = 0.5, 
    y = -log10(min(data$data.with.inflation$Probt)) - 1,
    labels = bquote(lambda[bacon] == .(format(data$estimatedInflation,digits = 4))),
    pos = 4
  )
}
```

### Males
```{R}
logistic_mixed_model_ad_cn.with.inflation <- estimation_of_inflation(data = logistic_mixed_model_ad_cn.annotated.males)
dim(logistic_mixed_model_ad_cn.with.inflation$data.with.inflation)
```

```{R, eval = F}
readr::write_csv(
  x = logistic_mixed_model_ad_cn.with.inflation$data.with.inflation,
  file = file.path(dir.results.glm,"Males_ADNI_AD_vs_CN_glmm_using_beta_with_inflation.csv")
)
```

```{R, fig.width = 6}
plot_inflation(
  data = logistic_mixed_model_ad_cn.with.inflation, 
  main = "MALES ADNI AD vs CN - single cpg\n Logistic regression model"
)

qqPlot(
  logistic_mixed_model_ad_cn.with.inflation$data.with.inflation$pValue.bacon, ci = TRUE, 
  main = paste0("MALES -  ADNI AD vs CN - single cpg\n logistic regression mode - pValue.bacon")
) 

```

### Females
```{R}
logistic_mixed_model_ad_cn.with.inflation <- estimation_of_inflation(data = logistic_mixed_model_ad_cn.annotated.females)
dim(logistic_mixed_model_ad_cn.with.inflation$data.with.inflation)
```

```{R, eval = F}
readr::write_csv(
  x = logistic_mixed_model_ad_cn.with.inflation$data.with.inflation,
  file = file.path(dir.results.glm,"Females_ADNI_AD_vs_CN_glmm_using_beta_with_inflation.csv")
)
```

```{R, fig.width = 6}
plot_inflation(
  data = logistic_mixed_model_ad_cn.with.inflation, 
  main = "FEMALE - ADNI AD vs CN - single cpg\n Logistic regression model"
)

qqPlot(
  logistic_mixed_model_ad_cn.with.inflation$data.with.inflation$pValue.bacon, ci = TRUE, 
  main = paste0("FEMALES ADNI AD vs CN - single cpg\n logistic regression model - pValue.bacon")
) 

```

# Session information
```{R}
devtools::session_info()
```